src/secret.rs

```diff
@@ -0,0 +1,185 @@
+/// let totp_plain = TOTP::new(
+///     Algorithm::SHA1,
+///     6,
+///     1,
+///     30,
+///     secret.as_bytes().unwrap(),
+///     None,
+///     "account".to_string(),
+/// ).unwrap();
+///
+/// println!("code from plain text:\t{}", totp_plain.generate_current().unwrap());
+/// ```
+///
+/// - Create a TOTP from a base32 encoded secret
+/// ```
+/// use totp_rs::{Secret, TOTP, Algorithm};
+///
+/// let secret = Secret::Base32(String::from("NV4S243FMNZGK5A"));
+/// let totp_base32 = TOTP::new(
+///     Algorithm::SHA1,
+///     6,
+///     1,
+///     30,
+///     secret.as_bytes().unwrap(),
+///     None,
+///     "account".to_string(),
+/// ).unwrap();
+///
+/// println!("code from base32:\t{}", totp_base32.generate_current().unwrap());
+///
+/// ```
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum Secret {
+    /// represent a non-encoded "plain text" secret
+    Plain(String),
+    /// represent a base32 encoded secret
+    Base32(String),
+}
+
+impl Secret {
+    /// Get the inner String value of the enum variant
+    pub fn inner(&self) -> &String {
+        match self {
+            Secret::Plain(s) => s,
+            Secret::Base32(s) => s,
+        }
+    }
+
+    /// Get the inner String value as a Vec of bytes
+    pub fn as_bytes(&self) -> Result<Vec<u8>, SecretParseError> {
+        match self {
+            Secret::Plain(s) => Ok(s.as_bytes().to_vec()),
+            Secret::Base32(s) => match base32::decode(Alphabet::RFC4648 { padding: false }, s) {
+                Some(bytes) => Ok(bytes),
+                None => Err(SecretParseError::ParseBase32),
+            },
+        }
+    }
+
+    /// Transforms a `Secret::Base32` into a `Secret::Plain`
+    pub fn as_plain(&self) -> Result<Self, SecretParseError> {
+        match self {
+            Secret::Plain(_) => Ok(self.clone()),
+            Secret::Base32(s) => match base32::decode(Alphabet::RFC4648 { padding: false }, s) {
+                Some(buf) => match String::from_utf8(buf) {
+                    Ok(str) => Ok(Secret::Plain(str)),
+                    Err(e) => Err(SecretParseError::Utf8Error(e)),
+                },
+                None => Err(SecretParseError::ParseBase32),
+            },
+        }
+    }
+
+    /// Transforms a `Secret::Plain` into a `Secret::Base32`
+    pub fn as_base32(&self) -> Self {
+        match self {
+            Secret::Plain(s) => Secret::Base32(base32::encode(
+                Alphabet::RFC4648 { padding: false },
+                s.as_ref(),
+            )),
+            Secret::Base32(_) => self.clone(),
+        }
+    }
+
+    /// ⚠️ requires feature `gen_secret`
+    ///
+    /// Generate a CSPRNG alpha-numeric string of length `size`
+    #[cfg(feature = "gen_secret")]
+    pub fn generate_secret(size: usize) -> Secret {
+        use rand::distributions::{Alphanumeric, DistString};
+        Secret::Plain(Alphanumeric.sample_string(&mut rand::thread_rng(), size))
+    }
```

Comment:
Why genererate a random string instead of u8[20] ?

Secrets, outside of already base32'd secrets, will generally not be strings. In the examples, I use strings for readability, but using the Alnum range (just under 6 bits per byte) seems to decrease the number of random bits by a lot?

Reply:
I could add a `Secret::Bytes(Vec<u8>)` variant and generate this by default.

The idea of using the `Alphanumeric` distribution was to ensure that the bytes generated are utf-8 values and can be converted to base32.

Looking at the [docs](https://docs.rs/rand/0.8.5/rand/distributions/index.html) of rand now I see that the `Standard` distribution can generate `char` which should be ok. I'd have to test this.

Reply:
IIRC any byte will be a printable char once base32'd, so I think that would be preferable

There's no rush to do it right now, but I'd appreciate it very much if you could test it (I'm leaving work in a few hours, I'll be able to look at it more in depth by then)

Reply:
From the docs, I get the impression the char distribution excludes some codepoint, so u8 would probably be better at the u8 Standard distribution will take advantage of the full range of value its type can hold

Reply:
`u8[20]` would have the disadvantage of generating non utf-8 sequence and not being convertible back to `Secret::Plain`

I did some testing for both:
```rust
use rand::Rng;

let mut rng = rand::thread_rng();
let mut s: String = String::new();
for _x in 0..=20 {
    let c: char = rng.gen();
    s = format!("{}{}", &s, c);
}
println!("{s}");
let enc = base32::encode(base32::Alphabet::RFC4648 { padding: false }, s.as_bytes());
println!("{enc}");
let secret = Secret::Base32(enc);
println!("{}", secret.as_plain().unwrap())
```

```rust
use rand::Rng;

let mut rng = rand::thread_rng();
let mut s: [u8; 20] = Default::default();
rng.fill(&mut s);
println!("{s:?}");
let enc = base32::encode(base32::Alphabet::RFC4648 { padding: false }, &s);
println!("{enc}");
let secret = Secret::Base32(enc);
println!("{}", secret.as_plain().unwrap());
```

the first one seems to run just fine, the second one pretty much always fails at `secret.as_plain()`

Also as a note the size recommendation of the rfc is in bits which was 20 `Alphanumeric` chars, but generating random `char` means the bit size is not constant anymore (8 to 32 bits per char if I'm not mistaken).
Might be a bit tricky to generate exactly 160 bits, we could settle for 160+ or 128+.

Reply:
`Plain` would be used as input for disambiguation, but I'm not sure they should be used as output, because keys are binary 160 keys, and thus not guaranteed to be valid ASCII, let alone UTF-8 sequences

Reply:
Maybe the Secret Enum should be  (so that it can accept a string easily

```Rust
pub enum Secret {
    /// represent a non-encoded "raw" secret
    Raw(&[u8]),
    /// represent a base32 encoded secret
    EncodedString),
}```